### PR TITLE
Download a bee photo

### PR DESCRIPTION
**Pull request by lawsonbh**
This PR:
1. Implements the /download_bee route which uses a function named download_file. It's similar in structure to how we implemented the /upload_bee_photo route. We can change this in the future to be an id value associated with each resource in the database.

I'd love specific feedback on guiding where the user downloads the object to. Right now - the file will download to the project's home directory. I think for me as an admin - this is fine. I'd like your opinion on what to name the parameter for changing the path this file gets downloaded to.

### CODE REVIEW

**Comment by lawsonbh** [comment]
And actually ... right now this isn't returning a dictionary with a key, value including the path the file was downloaded to. It returns that none type I mentioned in the #3 PR. I can create the f string like we did in the upload_file implementation but wanted to see if we could take a look at it together first. Screenshot of null attached.
null_type

File: main.py [hunk 1]
```diff
@@ -58,9 +58,43 @@ def upload_file(
...
+    aws_secret = os.environ["AWS_SECRET_ACCESS_KEY"]
+
+    s3_client = boto3.client(
+        "s3", aws_access_key_id=aws_key, aws_secret_access_key=aws_secret
+    )
+    # TODO: Create parameter in func for changing Key and Filename values
+    response = s3_client.download_file(
+        Bucket=s3_bucket, Filename=file_name, Key=file_name
+    )
+    return response
```
Comment (bbelderbos) [inline main.py:99]: I would return the url like we did with upload

File: main.py [hunk 2]
```diff
@@ -58,9 +58,43 @@ def upload_file(
...
+    """Upload a file like object to an S3 bucket
+    :param file_name: Name of the file to download including its extension
+    :param bucket: the S3 bucket to download the file from
+    :param acl: the S3 access control list type
+    :return: dict containing a value with the path to the file you downloaded
+    """
+    s3_bucket = bucket or os.environ["AWS_S3_BUCKET"]
+
+    aws_key = os.environ["AWS_ACCESS_KEY_ID"]
+    aws_secret = os.environ["AWS_SECRET_ACCESS_KEY"]
```
Comment (bbelderbos) [inline main.py:88]: we can turn the AWS S3 functions into a class later to remove some duplication, e.g. setting key + secret in the class constructor ...

File: main.py [hunk 3]
```diff
@@ -58,9 +58,43 @@ def upload_file(
...
+
+    s3_client = boto3.client(
+        "s3", aws_access_key_id=aws_key, aws_secret_access_key=aws_secret
+    )
+    # TODO: Create parameter in func for changing Key and Filename values
+    response = s3_client.download_file(
```
Comment (bbelderbos) [inline main.py:95]: what is the response if the file is not present?

File: main.py [hunk 4]
```diff
@@ -58,9 +58,43 @@ def upload_file(
...
+    bee_photo_name: str, bucket: Optional[str] = None, acl: Optional[str] = None
+) -> str:
+    """
+    Download a bee photo from an S3 bucket
+    - **bee_photo_name**: required, str including file extension
+    - **bucket**: the S3 bucket to post the file object to
+    - **acl**: the S3 access control list type
+    Returns a dictionary with the path of the file you downloaded
+    """
+    response = download_file(file_name=bee_photo_name, bucket=bucket, acl=acl)
```
Comment (bbelderbos) [inline main.py:72]: depending the response of the helper function we can also raise a FastAPI exception here if file is not found, e.g.

```
raise HTTPException(status_code=404, detail=f"{bee_photo_name} not found.")
```

**Review by bbelderbos — APPROVED** [review]
Good work, few suggestions.

HTH
Bob

File: main.py [hunk 5]
```diff
@@ -99,6 +99,71 @@ def download_file(
...
+    """List all objects in the S3 bucket. Hopefully its a bunch of bees in hats!
+    Returns a dictionary where the value of the message key is the list of object
+    names as strings.
+    """
+    response = list_objects()
+
+    return {"message": response}
+
+
+def list_objects() -> list:
```
Comment (bbelderbos) [inline main.py:113]: seems this returns a generator expression

File: main.py [hunk 6]
```diff
@@ -58,7 +58,7 @@ def upload_file(
     return response
 
 
-@app.get("/download_bee/")
+@app.get("/download_bee/", status_code=204)
```
Comment (bbelderbos) [inline main.py:61]: This seems a delete status code: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204

File: main.py (hunk 5 above)
Comment (bbelderbos) [reply main.py:113]: you can further type a list, e.g. `list[str]`

File: main.py [hunk 7]
```diff
@@ -99,6 +99,71 @@ def download_file(
...
+    return {"message": response}
+
+
+def list_objects() -> list:
+    """Set up access to the S3 bucket and call the generator to create a list of
+    all objects.
+    :return: a list of objects from the bucket
+    """
+
+    s3_bucket = os.environ["AWS_S3_BUCKET"]
```
Comment (bbelderbos) [inline main.py:119]: we can centralize this in a class constructor to deduplicate code, but that's for later

File: main.py [hunk 8]
```diff
@@ -99,6 +99,71 @@ def download_file(
...
+
+
+def generate_bucket_keys(
+    paginator,
+    bucket_name: str,
+    prefix: Optional[str] = "/",
+    delimiter: Optional[str] = "/",
+    start_after: Optional[str] = "",
+):
+    """Function to generate and yield all object names from an S3 bucket. Follows
```
Comment (bbelderbos) [inline main.py:145]: interesting

File: main.py [hunk 9]
```diff
@@ -99,6 +99,71 @@ def download_file(
...
+
+    all_bees = (
+        key
+        for key in generate_bucket_keys(paginator=s3_paginator, bucket_name=s3_bucket)
+    )
+
+    return all_bees
+
+
+def generate_bucket_keys(
```
Comment (bbelderbos) [inline main.py:138]: I like that you have different functions calling each other, more modular

File: main.py [hunk 10]
```diff
@@ -99,6 +99,71 @@ def download_file(
...
+    :param bucket_name: S3 bucket to list the contents of
+    :param prefix: S3 limits responses to keys that begin with this character or
+    set of characters
+    :param delimiter: a character or set of characters that groups objects in S3
+    :param start_after: where to start listing from, S3 starts listing after the
+    specified key
+    :return: generator yielding a string key to the object in the specifed S3
+    bucket
+    """
+    prefix = prefix[1:] if prefix.startswith(delimiter) else prefix
```
Comment (bbelderbos) [inline main.py:158]: can you explain this and next line, not sure if I fully understand ...

**Review by bbelderbos — COMMENTED** [review]
Good work!

**Pull request merged**